File: rust/anagrams.rs
use std::io::Write;

fn main() {
    let args = std::env::args();
    if args.len() != 2 {
        let _ = writeln!(std::io::stderr(), "Please enter exactly one argument");
        std::process::exit(1);
    }
    if let Some(last_arg) = args.last() {
        let mut word: Vec<_> = last_arg.chars().collect();
        generate_permutations(word.len() - 1, &mut word);
    }
}

fn generate_permutations(n: usize, a: &mut Vec<char>) {
    if n == 0 {
        println!("{}", a.clone().into_iter().collect::<String>());
    } else {
        for i in 0..n+1 {
            generate_permutations(n - 1, a);
            a.swap(if n % 2 == 0 { i } else { 0 }, n);
        }
    }
}

File: swift/anagrams.swift
import Foundation

func generatePermutations(inout a : [Character], upTo n: Int) {
    if n == 0 {
        print(String(a))
    } else {
        for i in 0..<n {
            generatePermutations(&a, upTo: n-1)
            swap(&a[n % 2 == 0 ? 0 : i], &a[n])
        }
        generatePermutations(&a, upTo: n-1)
    }
}

if Process.arguments.count != 2 {
    fputs("Exactly one argument is required\n", __stderrp)
    exit(1)
}
let word = Process.arguments[1]
var charArray = Array(word.characters)
generatePermutations(&charArray, upTo: charArray.count-1)
